File: scripts-rs/src/compiler/context.rs
use crate::core::Paths;
use crate::diagnostics::{Diagnostics, DiagnosticSource, Diagnostic};
use std::time::Instant;

#[derive(Clone, Debug)]
pub struct CompilerConfig {
    pub docs_dir: String,
    pub output_dir: String,
    pub site_url: String,
}

pub struct CompilationContext {
    pub config: CompilerConfig,
    pub diagnostics: Diagnostics,
    pub start_time: Instant,
}

impl CompilationContext {
    pub fn new(paths: &Paths) -> Self {
        Self::from_config(CompilerConfig {
            docs_dir: paths.root.join("docs").to_string_lossy().to_string(),
            output_dir: paths.root.join("src").join("generated").to_string_lossy().to_string(),
            site_url: "https://your-docs-site.com".to_string(), // TODO: config
        })
    }

    pub fn from_config(config: CompilerConfig) -> Self {
        Self {
            config,
            diagnostics: Diagnostics::new(),
            start_time: Instant::now(),
        }
    }

    pub fn report(&mut self, diag: Diagnostic) {
        self.diagnostics.report(diag);
    }

    pub fn error(&mut self, source: DiagnosticSource, file: &str, message: &str, detail: Option<&str>) {
        self.diagnostics.error(source, file, message, detail);
    }

    pub fn warn(&mut self, source: DiagnosticSource, file: &str, message: &str, detail: Option<&str>) {
        self.diagnostics.warn(source, file, message, detail);
    }

    pub fn info(&mut self, source: DiagnosticSource, file: &str, message: &str) {
        self.diagnostics.info(source, file, message);
    }
}

File: scripts/compiler/Context.ts
/**
 * Compilation Context
 * Manages shared state, diagnostics, and configuration during a build.
 */

import { colors } from "../core/index.ts";
import type { CompilerConfig, Diagnostic, DiagnosticSeverity, DiagnosticSource } from "./types.ts";

export class CompilationContext {
  private diagnostics: Diagnostic[] = [];
  public readonly config: CompilerConfig;
  public startTime: number = Date.now();

  constructor(config: CompilerConfig) {
    this.config = config;
  }

  report(diag: Diagnostic): void {
    this.diagnostics.push(diag);
  }

  error(
    source: DiagnosticSource,
    file: string,
    message: string,
    detail?: string,
    line?: number
  ): void {
    this.report({ severity: "error", source, file, message, detail, line });
  }

  warn(
    source: DiagnosticSource,
    file: string,
    message: string,
    detail?: string,
    line?: number
  ): void {
    this.report({ severity: "warning", source, file, message, detail, line });
  }

  info(source: DiagnosticSource, file: string, message: string): void {
    this.report({ severity: "info", source, file, message });
  }

  getDiagnostics(): ReadonlyArray<Diagnostic> {
    return this.diagnostics;
  }

  hasErrors(): boolean {
    return this.diagnostics.some((d) => d.severity === "error");
  }

  getSummary() {
    let errors = 0;
    let warnings = 0;
    let info = 0;
    for (const d of this.diagnostics) {
      if (d.severity === "error") errors++;
      else if (d.severity === "warning") warnings++;
      else info++;
    }
    return { errors, warnings, info };
  }

  formatReport(): string {
    if (this.diagnostics.length === 0) return `${colors.green}✓ No issues detected${colors.reset}`;

    const lines: string[] = [];
    const severityIcon: Record<DiagnosticSeverity, string> = {
      error: `${colors.red}✗${colors.reset}`,
      warning: `${colors.yellow}⚠${colors.reset}`,
      info: `${colors.blue}ℹ${colors.reset}`,
    };

    for (const d of this.diagnostics) {
      const icon = severityIcon[d.severity];
      const pos = d.line ? `:${d.line}` : "";
      const header = `${icon} ${colors.bright}[${d.severity.toUpperCase()}]${colors.reset} ${colors.cyan}${d.file}${colors.reset}${pos} (${colors.dim}${d.source}${colors.reset})`;
      lines.push(header);
      lines.push(`   ${d.message}`);
      if (d.detail) lines.push(`   ${colors.dim}→ ${d.detail}${colors.reset}`);
    }

    const { errors, warnings, info } = this.getSummary();
    lines.push("");
    lines.push(
      `${colors.bright}Summary:${colors.reset} ${colors.red}${errors} error(s)${colors.reset}, ${colors.yellow}${warnings} warning(s)${colors.reset}, ${colors.blue}${info} info${colors.reset}`
    );
    return lines.join("\n");
  }
}
